# MLOps Workshop: Modell-Engineering und Training

## Einführung
In diesem Notebook werden wir uns mit dem Modell-Engineering und Training für unser Customer Churn Prediction Projekt beschäftigen. Wir werden MLflow für das Experiment-Tracking verwenden und verschiedene Modelle evaluieren.

## Lernziele
Nach Abschluss dieses Notebooks werden Sie:
- Verstehen, wie man MLflow für Experiment-Tracking einsetzt
- Verschiedene Modelle mit unterschiedlichen Hyperparametern trainieren können
- Best Practices für Modell-Evaluierung kennen
- Wissen, wie man Modelle im MLflow Model Registry registriert

## 1. Setup und Daten laden


In [22]:
# Benötigte Bibliotheken importieren
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.ensemble import GradientBoostingClassifier
import mlflow
import mlflow.sklearn
import datetime
import matplotlib.pyplot as plt
import seaborn as sns

# MLflow Tracking lokal einrichten (ohne Server)
mlflow.set_tracking_uri("file:./mlruns")


In [23]:
# Daten laden und vorbereiten
processed_data = pd.read_csv('../data/processed/loan_data_preprocessed.csv')



In [19]:
# Train-Test-Split
X = processed_data.drop(columns="not.fully.paid")
y= processed_data["not.fully.paid"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train= scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [20]:
# Modell erstellen

model = LogisticRegression(max_iter=1000, class_weight="balanced")
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
y_prob = model.predict_proba(X_test)[:,1]

print("Accuracy:", accuracy_score(y_test, y_pred))
print("ROC-AUC:", roc_auc_score(y_test, y_prob))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))


Accuracy: 0.6482254697286013
ROC-AUC: 0.7127535081560176
[[1033  578]
 [  96  209]]
              precision    recall  f1-score   support

           0       0.91      0.64      0.75      1611
           1       0.27      0.69      0.38       305

    accuracy                           0.65      1916
   macro avg       0.59      0.66      0.57      1916
weighted avg       0.81      0.65      0.69      1916



In [26]:
import pandas as pd
import mlflow
import mlflow.sklearn
import datetime

from sklearn.base import clone
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from mlflow.tracking import MlflowClient

# MLflow sauber setzen
tracking_uri = "file:../mlruns"
experiment_name = "Loan Default Prediction"

mlflow.set_tracking_uri(tracking_uri)
client = MlflowClient(tracking_uri=tracking_uri)

# Falls das Experiment nur soft-deleted wurde, wiederherstellen
exp = client.get_experiment_by_name(experiment_name)
if exp is not None and exp.lifecycle_stage == "deleted":
    client.restore_experiment(exp.experiment_id)

mlflow.set_experiment(experiment_name)

# Daten laden
processed_data = pd.read_csv("../data/processed/loan_data_preprocessed.csv")

X = processed_data.drop(columns="not.fully.paid")
y = processed_data["not.fully.paid"]

# WICHTIG: stratify wegen Klassenungleichgewicht
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

def train_and_evaluate_model(model, X_train, X_test, y_train, y_test, run_name, params=None):
    with mlflow.start_run(run_name=run_name) as run:
        if params:
            mlflow.log_params(params)

        model.fit(X_train, y_train)

        y_pred = model.predict(X_test)

        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred, zero_division=0)
        recall = recall_score(y_test, y_pred, zero_division=0)
        f1 = f1_score(y_test, y_pred, zero_division=0)

        metrics = {
            "accuracy": accuracy,
            "precision": precision,
            "recall": recall,
            "f1_score": f1,
            "roc_auc": None
        }

        if hasattr(model, "predict_proba"):
            y_prob = model.predict_proba(X_test)[:, 1]
            metrics["roc_auc"] = roc_auc_score(y_test, y_prob)
            mlflow.log_metric("roc_auc", metrics["roc_auc"])

        mlflow.log_metric("accuracy", accuracy)
        mlflow.log_metric("precision", precision)
        mlflow.log_metric("recall", recall)
        mlflow.log_metric("f1_score", f1)

        mlflow.sklearn.log_model(model, name="model")

        return run.info.run_id, metrics


all_results = []

# 1) Logistic Regression mit Pipeline (Scaler nur hier)
logreg_params_list = [
    {"C": 0.1, "max_iter": 1000, "class_weight": None, "random_state": 42},
    {"C": 1.0, "max_iter": 1000, "class_weight": "balanced", "random_state": 42},
    {"C": 10.0, "max_iter": 1000, "class_weight": "balanced", "random_state": 42},
]

for params in logreg_params_list:
    model = Pipeline([
        ("scaler", StandardScaler()),
        ("model", LogisticRegression(**params))
    ])

    run_id, metrics = train_and_evaluate_model(
        model,
        X_train,
        X_test,
        y_train,
        y_test,
        run_name="LogisticRegression",
        params=params
    )

    all_results.append({
        "run_id": run_id,
        "params": params,
        "metrics": metrics,
        "model_type": "LogisticRegression"
    })


# 2) Random Forest ohne Scaler
rf_params_list = [
    {"n_estimators": 100, "max_depth": 10, "min_samples_split": 2, "random_state": 42},
    {"n_estimators": 200, "max_depth": 15, "min_samples_split": 5, "random_state": 42},
    {"n_estimators": 300, "max_depth": None, "min_samples_split": 10, "random_state": 42},
]

for params in rf_params_list:
    model = RandomForestClassifier(**params)

    run_id, metrics = train_and_evaluate_model(
        model,
        X_train,
        X_test,
        y_train,
        y_test,
        run_name="RandomForest",
        params=params
    )

    all_results.append({
        "run_id": run_id,
        "params": params,
        "metrics": metrics,
        "model_type": "RandomForest"
    })


# 3) Gradient Boosting ohne Scaler
gb_params_list = [
    {"n_estimators": 100, "learning_rate": 0.05, "random_state": 42},
    {"n_estimators": 200, "learning_rate": 0.05, "random_state": 42},
    {"n_estimators": 100, "learning_rate": 0.1, "random_state": 42},
]

for params in gb_params_list:
    model = GradientBoostingClassifier(**params)

    run_id, metrics = train_and_evaluate_model(
        model,
        X_train,
        X_test,
        y_train,
        y_test,
        run_name="GradientBoosting",
        params=params
    )

    all_results.append({
        "run_id": run_id,
        "params": params,
        "metrics": metrics,
        "model_type": "GradientBoosting"
    })


# Bestes Modell wirklich aus ALLEN Modellen auswählen
best_run = max(
    all_results,
    key=lambda x: x["metrics"]["roc_auc"] if x["metrics"]["roc_auc"] is not None else x["metrics"]["f1_score"]
)

best_run_id = best_run["run_id"]
model_name = "loan_default_predictor"

model_version = mlflow.register_model(
    f"runs:/{best_run_id}/model",
    model_name
)

print(f"Model '{model_name}' wurde als Version {model_version.version} registriert")
print("Bestes Modell:")
print(best_run["model_type"])
print(best_run["params"])
print(best_run["metrics"])


# Modell auf Staging setzen und Beschreibung aktualisieren
client = MlflowClient()

client.transition_model_version_stage(
    name=model_name,
    version=model_version.version,
    stage="Staging",
    archive_existing_versions=True
)

client.update_model_version(
    name=model_name,
    version=model_version.version,
    description=f"""
Loan Default Prediction Model v{model_version.version}

- Modelltyp: {best_run['model_type']}
- Training durchgeführt am: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M')}
- Beste Parameter: {best_run['params']}

Performance:
- Accuracy: {best_run['metrics']['accuracy']:.4f}
- Precision: {best_run['metrics']['precision']:.4f}
- Recall: {best_run['metrics']['recall']:.4f}
- F1-Score: {best_run['metrics']['f1_score']:.4f}
- ROC-AUC: {best_run['metrics']['roc_auc']:.4f}
"""
)

print("Modell erfolgreich registriert und auf Staging gesetzt.")

2026/03/11 16:05:15 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/11 16:05:19 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/11 16:05:23 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mec

Model 'loan_default_predictor' wurde als Version 9 registriert
Bestes Modell:
LogisticRegression
{'C': 1.0, 'max_iter': 1000, 'class_weight': 'balanced', 'random_state': 42}
{'accuracy': 0.6388308977035491, 'precision': 0.24091520861372812, 'recall': 0.5830618892508144, 'f1_score': 0.34095238095238095, 'roc_auc': 0.6803991392067827}
Modell erfolgreich registriert und auf Staging gesetzt.


### Auswertung der Modellleistung

Das trainierte Modell erreicht eine Accuracy von 0.65 sowie eine ROC-AUC von 0.71. Die ROC-AUC zeigt, dass das Modell grundsätzlich in der Lage ist, zwischen vollständig zurückgezahlten Krediten und nicht vollständig zurückgezahlten Krediten zu unterscheiden. Ein Wert von über 0.7 deutet auf eine moderate Trennfähigkeit des Modells hin.

Die Confusion Matrix zeigt, dass von insgesamt 305 tatsächlich nicht vollständig zurückgezahlten Krediten 209 korrekt erkannt wurden, während 96 Ausfälle nicht erkannt wurden. Daraus ergibt sich ein Recall von 0.69 für die Klasse „not.fully.paid = 1“, was bedeutet, dass etwa **69 % der tatsächlichen Kreditausfälle identifiziert werden.

Die Precision für die Ausfallklasse beträgt 0.27. Das bedeutet, dass nur etwa 27 % der als riskant eingestuften Kredite tatsächlich nicht vollständig zurückgezahlt werden, während ein größerer Teil dieser Vorhersagen Fehlalarme darstellt. Dies wird auch in der Confusion Matrix sichtbar, in der 578 Kredite fälschlicherweise als riskant klassifiziert wurden.

Insgesamt zeigt das Modell damit eine relativ hohe Sensitivität gegenüber Kreditausfällen, erkennt also einen großen Teil der tatsächlich problematischen Kredite. Gleichzeitig entstehen jedoch viele False Positives, da das Modell auch Kredite als riskant einstuft, die letztlich vollständig zurückgezahlt werden.

Diese Ergebnisse verdeutlichen, dass das Modell grundsätzlich Muster im Datensatz erkennen kann, die mit Kreditausfällen zusammenhängen, die Vorhersagen jedoch weiterhin mit einer gewissen Unsicherheit verbunden sind.

## 2. MLflow Experiment erstellen

### Aufgabe 1
Erstellen Sie ein neues MLflow Experiment für unser Churn-Vorhersage-Projekt. Setzen Sie einen aussagekräftigen Namen und eine Beschreibung.

<details>
<summary>👉 Lösung anzeigen</summary>

```python
# MLflow Experiment erstellen
experiment_name = "customer_churn_prediction"
experiment = mlflow.set_experiment(experiment_name)

# Experiment-Details ausgeben
print(f"Experiment Name: {experiment_name}")
print(f"Experiment ID: {experiment.experiment_id}")
```
</details>